audit: Reduce the number of memory allocations in the hotpath

graphcareful opened the pull request:
With auditing enabled, high traffic events and low batch intervals it was observed that auditing was using more of the CPU then expected. This was traced back to many string conversions / string interpolation events when constructing an audit event.

An audit event is constructed at every possible attempt to enqueue the event so a hash code can be calculated. In the event the message has been determined it is a duplicate, the event can be deallocated, but this is wasteful. The event should only be constructed if it unique, to avoid the expensive serialization performed.

This pull request modifies the control flow in the audit log manager to calculate a hash code for the ocsf event without constructing the event itself. Since event construction depends on these parameters anyway, an appropriately unique hash code can be calculated with a free function using these parameters as input. In the event the hash code is unique, then the event will be allocated.

Linking to issue: https://github.com/redpanda-data/redpanda/issues/15898

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.3.x
- [ ] v23.2.x
- [ ] v23.1.x

## Release Notes

### Improvements

* Reduces the number of allocations performed by the auditing subsystem

graphcareful commented on src/v/security/audit/schemas/hashing_utils.cc:
```diff
@@ -0,0 +1,237 @@
+  ss::httpd::const_req req,
+  const ss::sstring& user,
+  const ss::sstring& svc_name) {
+    const auto activity_id = http_method_to_activity_id(req._method);
```
@michael-redpanda do we need the activity id to calculate the base hash? All that would matter is hashing the method I think. Then we could make this method private to utils.h again

michael-redpanda replied:
This I believe is for the API activity, yes?  API activity includes both HTTP and Kafka.  Kafka doesn't have a 'method' field.  So yes, I believe we do need to use activity ID in the hash.  You should double check, but does "type_id" of the base type get hashed?  If so then "activity_id" is already being hashed as Type ID is a combination of class ID and activity ID (https://schema.ocsf.io/1.0.0/classes/api_activity?extensions= look towards the bottom).

graphcareful replied:
Yeah for sure, i was just wondering if it would be good enough to has something like 0 - 5, one value for each of the potential HTTP method types, instead of having to get one of 5 stringified versions of the enum to later just hash

graphcareful commented:
I think it might also be worth it adding an additional find/replace commit to change the name of the `equality_fields()` family of methods. Currently they are now only used in the `estimated_ocsf_size()` method, I propose either changing the name of the method, or a small refactor to just have all ocsf types implement their own `size()` method (credit to Noah's original idea)

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/721cc98234209989fe3d0c6fcc487ec32e682ee3..acabea14001236c25cb3fbfc30bcd8be4d526d15)
- Fixed use after move
- Fixed bug with not accounting for size sempahore units

graphcareful commented on src/v/security/audit/audit_log_manager.h:
```diff
@@ -250,10 +226,50 @@ class audit_log_manager
-      std::unique_ptr<security::audit::ocsf_base_impl> msg);
     void set_enabled_events();
 
+    template<typename T, typename... Args>
```
Can use a concept here for T

StephanDollberg reviewed (commented):
Thanks for looking into this.

Had a look and approach seems good. It seems like we are hashing quite a bit more now but maybe this is just an impression. Hashing was already quite visible before so want to make sure we are as smart as possible there.

Will leave to enterprise for a more thorough review.

graphcareful commented:
> Thanks for looking into this.
> 
> Had a look and approach seems good. It seems like we are hashing quite a bit more now but maybe this is just an impression. Hashing was already quite visible before so want to make sure we are as smart as possible there.
> 
> Will leave to enterprise for a more thorough review.

Thanks for taking a look! We should be hashing the same amount as in the current impl since the hash function is called at the top of `do_enqueue_event` just like it stands as of now in dev.

BenPope commented on src/v/security/audit/schemas/hashing_utils.h:
```diff
@@ -0,0 +1,25 @@
+#include "security/audit/schemas/types.h"
+#include "security/authorizer.h"
+
+namespace security::audit {
```
The implementation might be a lot simpler with some helpers for hashing, perhaps something like:
```c++
namespace impl {
template<typename T>
concept is_absl_hashable = requires(T const& t) {
    { absl::Hash<T>{}(t) } -> std::convertible_to<std::size_t>;
};

template<typename T>
concept is_std_hashable = requires(T const& t) {
    { std::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
};

template<typename T>
concept is_boost_hashable = requires(T const& t) {
    { boost::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
};

template<typename T>
size_t hash_combine(size_t& seed, T const& t) {
    if constexpr (is_std_hashable<T>) {
        boost::hash_combine(seed, std::hash<T>{}(t));
    } else if constexpr (is_absl_hashable<T>) {
        boost::hash_combine(seed, absl::Hash<T>{}(t));
    } else if constexpr (is_boost_hashable<T>) {
        boost::hash_combine(seed, boost::hash<T>{}(t));
    } else {
        static_assert(always_false_v<T>, "No hasher found for T");
    }
    return seed;
}

} // namespace impl

template<typename... Args>
size_t hash_combine(size_t& seed, Args&&... args) {
    return (impl::hash_combine(seed, std::forward<Args>(args)), ...);
}
```
This would allow something like:
```diff
-    boost::hash_combine(h, std::hash<category_uid>()(ct_uid));
-    boost::hash_combine(h, std::hash<class_uid>()(c_uid));
-    boost::hash_combine(h, std::hash<severity_id>()(s_id));
+    hash_combine(h, ct_uid, c_uid, s_id);
```
or:
```diff
-        if (result.acl) {
-            boost::hash_combine(
-              h, absl::Hash<security::acl_entry>{}(*result.acl));
-        }
-        if (result.resource_pattern) {
-            boost::hash_combine(
-              h,
-              absl::Hash<security::resource_pattern>{}(
-                *result.resource_pattern));
-        }
+        hash_combine(h, result.acl, result.resource_pattern);
```

graphcareful replied:
Nice idea 👍

BenPope commented on src/v/security/audit/audit_log_manager.h:
```diff
@@ -232,10 +226,50 @@ class audit_log_manager
+    template<typename T, typename... Args>
+    bool do_enqueue_audit_event(Args&&... args) {
+        auto& map = _queue.get<underlying_unordered_map>();
+        const auto hash_key = T::hash(args...);
```
Would it not make sense to reuse an existing hashing extension point such as `std::hash`, `boost::hash`, or `absl::Hash`, rather than inventing another one?

graphcareful replied:
The difference here is that hash is a static method, we are calculating the hash at a point where the object doesn't yet exist.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/acabea14001236c25cb3fbfc30bcd8be4d526d15..ee41562cf993c6550f0c2567bc1abaff4cb9ea96)
- Added concept to constrain type T passed to the`do_enqueue` method
- Implemented Bens feedback making the hashing code more graceful
- Fixed a bug in a hash method that returned a value of 0 that attributed to a few ducktape test failures

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,71 @@
+    return seed;
+}
+
+template<typename T>
+size_t hash_combine(size_t& seed, std::optional<T> const& t) {
+    if (!t.has_value()) {
+        // Arbitray value for empty optional, taken from absl::hash
+        boost::hash_combine(seed, 0x12345678);
+        return seed;
+    }
+    return hash_combine<T>(seed, *t);
+}
```
Isn't `std::optional` already specialized for std, boost, and absl?

graphcareful replied:
Not for all of them

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,71 @@
+}
+
+template<typename... Args>
+size_t hash_combine(Args&&... args) {
```
I don't like this overload as it's a bit too easy to drop the seed, which would effectively ignore all the values before it,

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
```
nit: year

BenPope commented on src/v/security/audit/schemas/hashing_utils.cc:
```diff
@@ -0,0 +1,171 @@
+
+    // policy
+    enum class auth_state { disabled = 0, superuser = 1, nomatches = 2 };
+    using enum_t = std::underlying_type_t<auth_state>;
+    if (result.authorization_disabled) {
+        hash_combine(h, static_cast<enum_t>(auth_state::disabled));
+    } else if (result.is_superuser) {
+        hash_combine(h, static_cast<enum_t>(auth_state::superuser));
+    } else if (result.empty_matches) {
+        hash_combine(h, static_cast<enum_t>(auth_state::nomatches));
+    } else {
+        hash_combine(h, result.acl);
+        hash_combine(h, result.resource_pattern);
```
```suggestion
    if (result.authorization_disabled) {
        hash_combine(h, auth_state::disabled);
    } else if (result.is_superuser) {
        hash_combine(h, auth_state::superuser);
    } else if (result.empty_matches) {
        hash_combine(h, auth_state::nomatches);
    } else {
        hash_combine(h, result.acl, result.resource_pattern);
```

BenPope commented on src/v/security/audit/schemas/hashing_utils.cc:
```diff
@@ -0,0 +1,171 @@
+size_t http_request_hash(const ss::http::request& req) {
+    size_t h = 0;
+    /// request headers
+    for (const auto& hdr : req._headers) {
+        hash_combine(h, hdr.first);
+        hash_combine(h, hdr.second);
+    }
```
Is there a concern if the headers appear in a different order? I.e., should they be sorted or otherwise normalised?

Also:
```suggestion
    hash_combine(h, req._headers);
```

graphcareful replied:
no order doesn't matter

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,71 @@
+
+#include <functional>
+
+namespace impl {
```
I think it's worth having a namespace around all of this, so as not to pollute the global namespace.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ee41562cf993c6550f0c2567bc1abaff4cb9ea96..28b117d1899a6ffd9a291825e2317a8d7dbfdaf6)
- Fix for bug that included product in the application lifecycle msg unconditionally, making a dt test fail

graphcareful commented:
Change in [force-push](https://github.com/redpanda-data/redpanda/compare/28b117d1899a6ffd9a291825e2317a8d7dbfdaf6..c38c3d65dc615cdab8d610dfd9bbe27670388912)
- Implemented Bens updated feedback on namespace around the new hashing utils and modifications to calls to hash_combine

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c38c3d65dc615cdab8d610dfd9bbe27670388912..bed02c98b446ce18c20c40a25083dc3a095f11e5)
- Rebased against dev

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/43691#018cfec8-1b3f-4b54-a5d6-8864b63b3835:
```
"rptest.tests.transactions_test.TransactionsAuthorizationTest.simple_authz_test"
```
<!--0f3fe7bc-4af2-4aef-8571-c7a166e0e280-->

dotnwat commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,69 @@
+
+#pragma once
+
+#include "utils/functional.h"
```
how do we avoid making v::utils a dependency of v::hashing? even though this is header-only, it's going to break when we flip the switch later where the dependencies have to be explicitly recorded.

public headers should go into hashing/include/hashing

graphcareful replied:
yup will just copy that , removing the include

BenPope commented on src/v/security/acl.h:
```diff
@@ -375,6 +375,20 @@ class resource_pattern_filter
     const std::optional<ss::sstring>& name() const { return _name; }
     std::optional<pattern_filter_type> pattern() const { return _pattern; }
 
+    template<typename H>
+    friend H AbslHashValue(H h, const resource_pattern_filter& f) {
+        int8_t pattern = -2;
+        if (f._pattern) {
+            if (std::holds_alternative<pattern_match>(*f._pattern)) {
+                pattern = -1;
+            } else {
+                pattern = static_cast<std::underlying_type_t<pattern_type>>(
+                  std::get<pattern_type>(*f._pattern));
+            }
+        }
+        return H::combine(std::move(h), f._resource, f._name, pattern);
+    }
```
nit:
```suggestion
    template<typename H>
    friend H AbslHashValue(H h, const pattern_filter_type& f) {
        using pattern_t = std::underlying_type_t<pattern_type>;
        auto pattern = ss::visit(
          f,
          [](pattern_match const&) { return pattern_t{-1}; },
          [](pattern_type const& p) { return static_cast<pattern_t>(p); });
        return H::combine(std::move(h), pattern);
    }

    template<typename H>
    friend H AbslHashValue(H h, const resource_pattern_filter& f) {
        return H::combine(std::move(h), f._resource, f._name, f._pattern);
    }
```

BenPope replied:
You could also do this:
```suggestion
    template<typename H>
    friend H AbslHashValue(H h, const pattern_match&) {
        return H::combine(std::move(h), 0x1B3A5CD7); // random number
    }

    template<typename H>
    friend H AbslHashValue(H h, const resource_pattern_filter& f) {
        return H::combine(std::move(h), f._resource, f._name, f._pattern);
    }
```

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,69 @@
+    return seed;
+}
+
+template<typename T>
+size_t combine(size_t& seed, std::optional<T> const& t) {
+    if (!t.has_value()) {
+        // Arbitray value for empty optional, taken from absl::hash
+        boost::hash_combine(seed, 0x12345678);
+        return seed;
+    }
+    return combine<T>(seed, *t);
+}
```
This isn't needed (at least for clang/libc++)

graphcareful replied:
IMO keeping it is fine as logic stays the same. Otherwise will need to add conditional logic to only execute when T is a specialization of optional and the type is hashable by boost, is_std_optional is within `v::reflection`, don't want to have hashing depend on that either

BenPope replied:
Isn't this sufficient? https://en.cppreference.com/w/cpp/utility/optional/hash

graphcareful replied:
Removed method entirely, thanks Ben

BenPope commented on src/v/hashing/utils.h:
```diff
@@ -0,0 +1,62 @@
+
+namespace hash {
+namespace impl {
+/// Pulled from utils/functional.h to prevent making v::utils a dependency of
+/// v::hashing
+template<typename>
+inline constexpr bool always_false_v = false;
```
We don't need another instance of this, it should be the same as in `v::utils`, i.e., in the global namespace.
